### PR TITLE
Static check for zero_alloc: ignore allocation that lead to exceptional return

### PR DESCRIPTION
**Pull request by gretay-js**
A new version of #864 + #863 + #870. 

Best reviewed commit by commit. 

I can try to split them into separate PRs, but it is quite tricky to get the separate PRs to build and pass the tests on their own. 

- The main change: abstract domain is a triple `{ nor; exn; div; }` where each of the components represents different paths through the function: paths to normal return, paths to exception return, and diverging paths that never retrun, resp. Each component can have one of the following values: `Bot` for unreachable (empty set of paths), `Safe` for "guaranteed to not allocate on any of the relevant paths", and `Top` for arbitray allocation on the relevant paths. 
The "strict" check requires no allocation on any path. The "relaxed" check, which is the defaul, allows allocation on paths that lead to exceptional return from the function (i.e., `exn` can be `Top` in the summary of the function), but not on normal return or infinite loops. Allocatins on paths that end in a call to an external function that never returns, such as `exit`, are not ignored in "relaxed" mode (it was simpler to implement). 

- `Checkmach` uses `backend/dataflow.ml` which is essentially a copy of the upstream backwards dataflow analysis. The `div` component is computed as a postprocessing step using another run of the same dataflow analysis, after the other components are computed. To make this postprocessing faster and simpler to implement, there is a small change to the interface of `Dataflow` to expose the mapping of labels to abstract value.
 
- Attrbutes are now named `[@zero_alloc]` and `[@zero_alloc strict]` with the corresponding `[@zero_alloc assume]` and `[@zero_alloc strict assume]`.  The internal representation of these attributes has been reworked to make it easier to parse them and extend the static checker with other properties in the future. This involves changes throughout the compilation pipeline to propagate the new representation of these attributes all the way to the backend. Currently, this is in 3 separate separate commits. I can split this out as a separate PR, but it would mean fixing the existing `checkmach.ml`, which is substantially modified in the following steps.

- Better user-error messages: locations when check fails point to the attribute instead of the entire body of the function. `Location.t` is passed all the way to the backend. 

- Refactor cmx format representation of the results of the static checker: move `Checks` module out of `Cmx_format` into a separate module. Decomple the representation of function summaries used in `Compilenv` from the "raw" version saved to cmx. Change the "raw" representation from maps to lists. Should it use arrays? It is now easy to change. 

- Flambda2: results may be unsound in the presence of Delayed Try-with that cannot be represented as a Regular Try-with. I can make the analysis conservatively sound, but it causes the CI tests to fail, and there is no easy way to distinguish flambda2 in dune tests. Also I am not sure if the current optimizatons in flambda2 generate the cases. We need to fix `backend/dataflow.ml` to support them and maybe update `backend/liveness.ml` to use this dataflow as well. Separate PRs.

- There are several ways in which the current analysis can be optimized to use less memory and finish early when a function is established to be allocating. I am leaving them for later as it makes it harder to reason about soundness.

### CODE REVIEW

File: backend/printcmm.ml [hunk 1]
```diff
@@ -361,15 +361,18 @@ and sequence ppf = function
...
 
 let codegen_option = function
   | Reduce_code_size -> "reduce_code_size"
   | No_CSE -> "no_cse"
   | Use_linscan_regalloc -> "linscan"
-  | Assert p -> "assert "^(property p)
-  | Assume p -> "assume "^(property p)
+  | Check { property; strict; assume; loc = _ } ->
+    Printf.sprintf "%s %s%s"
+      (if assume then "assume" else "assert ")
```
Comment (xclerc) [inline backend/printcmm.ml:373]:
```suggestion
      (if assume then "assume" else "assert")
```

File: middle_end/flambda2/terms/check_attribute.ml [hunk 2]
```diff
@@ -22,26 +22,33 @@ end
...
+        loc : Location.t
+      }
 
 let print ppf t =
   match t with
   | Default_check -> ()
-  | Assert p -> Format.fprintf ppf "@[assert %s@]" (Property.to_string p)
-  | Assume p -> Format.fprintf ppf "@[assume %s@]" (Property.to_string p)
+  | Check { property; strict; assume; loc = _ } ->
+    Format.fprintf ppf "@[%s %s %a@]"
```
Comment (xclerc) [inline middle_end/flambda2/terms/check_attribute.ml:36]:
```suggestion
    Format.fprintf ppf "@[%s%s %a@]"
```

File: ocaml/lambda/lambda.mli [hunk 3]
```diff
@@ -310,8 +310,17 @@ type poll_attribute =
...
 type check_attribute =
   | Default_check
-  | Assert of property
-  | Assume of property
+  | Check of { property: property;
+               strict: bool;
+               (* [strict=true] property holds on all paths.
+                  [strict=false] if the function returns normally or diverges,
+                  then the property holds (but property violations on
+                  exceptional returns are ignored. *)
```
Comment (xclerc) [inline ocaml/lambda/lambda.mli:318]:
```suggestion
                  exceptional returns are ignored). *)
```

File: backend/checkmach.ml [hunk 4]
```diff
@@ -26,455 +26,726 @@
...
-    val add_value : Format.formatter -> t -> string -> Value.t -> unit
+  val find_opt : t -> string -> Func_info.t option
 
-    val add_dep : t -> callee:string -> caller:string -> unit
+  (** [join_value t name v] name must be in the current compilation unit,
+      and previously recorded.  *)
+  val join_value : t -> string -> Value.t -> unit
 
-    val in_current_unit : t -> string -> Debuginfo.t -> unit
+  (** [join_value t name v] name must be in the current compilation unit
```
Comment (xclerc) [inline backend/checkmach.ml:320]:
```suggestion
  (** [add_value t name v] name must be in the current compilation unit
```

**Comment by xclerc** [comment]
I am not sure I understand the semantics of
the non-strict variant. My understanding is
that it would accept allocations on paths
leading to an exception, but the following is
being rejected:
```
exception E of string
let[@zero_alloc] fail () = raise (E "msg")
```

**Comment by gretay-js** [comment]
This example is accepted when we pass `-g` compiler flag. Without this flag, `raise` is treated as `raise_notrace`, which has a special case in the analysis. If we don't treat `raise_notrace` specially, we will end up accepting the following example: 

```
exception Found

let[@zero_alloc strict] find a =
  let i = 0 in
  while true do
    if a.(i) then
      raise_notrace Found;
  done

let[@zero_alloc] foo n =
  let a = Array.make n true in
  find a
```
The dependency on a seemingly unrelated compiler flag is quite annoying for testing, but allocations depend on compiler configuration and optimization flags.

**Comment by xclerc** [comment]
Noted; here is another example I don't think I understand
(failing while I am expecting it to pass):
```
let[@inline never] forever () = while true do () done
let[@zero_alloc strict] f g x =
  try
    while true do
      forever ()
    done;
    x
  with _ ->
    g x;
    x
```
It works if `never` is changed to `always`. Isn't the `div`
component expected to provide support for such cases?

**Comment by xclerc** [comment]
I am not sure whether it is related to the previous
comment or is more linked to how `callees` and
`callers` work, but the following is accepted:

```
let[@zero_alloc] rec forever () = while true do () done
and[@zero_alloc] order () =
  forever ();
  alloc (2, 3)
```

while the following is rejected:

```
let[@zero_alloc] rec order () =
  forever ();
  alloc (2, 3)
and[@zero_alloc] forever () = while true do () done

```

**Comment by xclerc** [comment]
Back to https://github.com/ocaml-flambda/flambda-backend/pull/1157#issuecomment-1453236593, it seems to fail if
`raise ...` is replaced with `failwith "..."`
even though `stdlib` is apparently built with
`-g`.

**Comment by lthls** [comment]
I can write the code for correctly handling delayed try-with in `dataflow.ml` if you want, it shouldn't take me too long.

Regarding this PR, the problem is quite tricky and I think this PR comes short in a few places.
First, the handling of future functions is not perfect. The ideal solution would be a fixpoint computation, but that may end up too expensive. The current solution should be correct but could give false positives.
One example where it could matter:
```ocaml
exception E
let rec f x b =
  if b then (g x; raise E)
  else ()
and g y = ignore (y, y) (* Allocating *)
```
With a fixpoint computation, we should be able to infer that `f` only allocates on exceptional paths, but since we haven't seen `g` yet we can only record a dependency. When `g` gets treated later we find that it allocates, and propagate the property to its caller `f`, without the information that it's only on exceptional paths.

More problematic, the handling of exception paths looks weird. I think this is because the allocation check is naturally expressed as a forward dataflow analysis, but you're trying to use the backwards one. It should be possible to make it work anyway, but it doesn't look obvious when looking at your code that it is correct (the fine details around the handling of `raise` look like a red flag).
I would suggest to rewrite your analysis either without using `Dataflow`, or with a new `Dataflow.Forward` functor (I can write this one if you need it).

Divergence can be handled relatively easily in both forward and backward analyses. For the backwards one, I think it can be done by specifying a `?(diverging = D.bot)` parameter to `analyze` and use that in `get_lbl` instead of `D.bot` for the `Not_found` case. In your case, since you want to handle diverging paths in the same way as normal paths, I think you could use `Value.normal_return` as the diverging value (and get rid of the `div` field of your record).

File: backend/checkmach.ml [hunk 5]
```diff
@@ -26,455 +26,726 @@
...
+
+  val relaxed : t
+
+  val never_returns : t -> bool
+
+  val print : Format.formatter -> t -> unit
+
+  val transform : t -> t
+end = struct
+  (** Lifts V to pairs  *)
```
Comment (xclerc) [inline backend/checkmach.ml:129]:
```suggestion
  (** Lifts V to triplets  *)
```

File: backend/checkmach.ml [hunk 6]
```diff
@@ -26,455 +26,726 @@
...
+    | Invalid { a; fun_name; property } ->
+      Some
+        (Location.error_of_printer ~loc:a.loc
+           (print_error ~fun_name ~property)
+           a)
+    | _ -> None
 end
 
 module Func_info = struct
+  (* Fields are mutable: [annotation] change when a call appears before the
```
Comment (xclerc) [inline backend/checkmach.ml:262]:
```suggestion
  (* Fields are mutable: [annotation] changes when a call appears before the
```

File: backend/checkmach.ml [hunk 7]
```diff
@@ -26,455 +26,726 @@
...
-  val add_callee : string -> Cmx_format.checks -> unit
+  (** [get_value f] returns the value recorded for function [f] in [Compenv], either
+      because the check passed or because of user-defined "assume" annotation. *)
+  val get_value : string -> Value.t
 
-  (** Does the callee pass the check (i.e., satisfies the conditions such as no
-      allocation)? Returns [true] if the callee passed the check or the callee
-      is annotated to be skipped by this check. *)
-  val check_callee : string -> Cmx_format.checks -> bool
+  (** [set_value f v] record the value of the function named [f] in [Compenv]. *)
```
Comment (xclerc) [inline backend/checkmach.ml:292]:
```suggestion
  (** [set_value f v] record the value of the function named [f] in [Compilenv]. *)
```

File: backend/checkmach.ml [hunk 8]
```diff
@@ -26,455 +26,726 @@
...
-
-  let add_callee s (checks : Cmx_format.checks) =
-    if not (check_callee s checks)
-    then
-      checks.ui_noalloc_functions
-        <- String.Set.add s checks.ui_noalloc_functions
-
-  let check_specific s = not (Arch.operation_allocates s)
+  (* Compact the mapping from function name to Value.t to reduce size of Checks
+     in cmx and memory consumption Compilev. Different component have different
```
Comment (xclerc) [inline backend/checkmach.ml:683]:
```suggestion
     in cmx and memory consumption Compilenv. Different components have different
```

**Comment by gretay-js** [comment]
> Back to [#1157 (comment)](https://github.com/ocaml-flambda/flambda-backend/pull/1157#issuecomment-1453236593), it seems to fail if `raise ...` is replaced with `failwith "..."` even though `stdlib` is apparently built with `-g`.

You are right, there is a problem with `stdlib.cmx`, it has a custom build rule (hardcoded in dune) and does not get compiled with -alloc-check. I'll fix it separately... somehow. 

The use of `failwith str`  satisfies the annotation `[@zero_alloc]` with flambda and flambda2, and fails with closure, where `failwith` does not seem to be inlined. I added a test with flambda.  
Special case `failwith "string constant"` is fine because the exception gets statically allocated. Something like `failwith ("my message " ^str)` fails with flambda because ( ^ ) has no summary.

**Comment by gretay-js** [comment]
@xclerc  there were a few bug in handling of loops, that's why the examples you tried didn't work as expected. Thanks for pointing out these problems! I fixed them and added tests for loops.

**Comment by gretay-js** [comment]
> I can write the code for correctly handling delayed try-with in `dataflow.ml` if you want, it shouldn't take me too long.

@lthls 
great, thank you! I'll review the PR tomorrow. 


> Regarding this PR, the problem is quite tricky and I think this PR comes short in a few places. First, the handling of future functions is not perfect. The ideal solution would be a fixpoint computation, but that may end up too expensive. The current solution should be correct but could give false positives. One example where it could matter:
> 
> ```ocaml
> exception E
> let rec f x b =
>   if b then (g x; raise E)
>   else ()
> and g y = ignore (y, y) (* Allocating *)
> ```
> 
> With a fixpoint computation, we should be able to infer that `f` only allocates on exceptional paths, but since we haven't seen `g` yet we can only record a dependency. When `g` gets treated later we find that it allocates, and propagate the property to its caller `f`, without the information that it's only on exceptional paths.

Yes, the analysis is too conservative at the moment for these cases. I should have been more explicit about this case in the description.  The current PR can be useful in many places as it is.  I would like to merge this basic functionality in (as long as there are no soundness bugs) and then implement the fixpoint computation as a follow up PR. 
I think it can be done without keeping Mach of all unresolved functions around until the end of the compilation unit, using a symbolic representation of dependencies, and computing fixpoint on these symbolic summaries, but maybe it's too complicated.

> 
> More problematic, the handling of exception paths looks weird. I think this is because the allocation check is naturally expressed as a forward dataflow analysis, but you're trying to use the backwards one. 

The reason I switched to using Backward analysis is to compute backward reachability from normal and exceptional return.

> It should be possible to make it work anyway, but it doesn't look obvious when looking at your code that it is correct (the fine details around the handling of `raise` look like a red flag). 

Are you referring to the special handling of raise_notrace using `Value.join exn Value.safe` ? It is safe but may be too conservative. 

> I would suggest to rewrite your analysis either without using `Dataflow`, or with a new `Dataflow.Forward` functor (I can write this one if you need it).

I think we would still need a backward analysis first, then another pass as forward analysis to compute allocations that can reach each program point, and then a third pass to combine the results. It makes "future functions" harder to handle. 

> Divergence can be handled relatively easily in both forward and backward analyses. For the backwards one, I think it can be done by specifying a `?(diverging = D.bot)` parameter to `analyze` and use that in `get_lbl` instead of `D.bot` for the `Not_found` case. 

I made these changes, thanks for the suggestion.  It also reveal a bug in my previous handling of loops (The hashtable Dataflow.lbls does not include labels whose value is  Bot).

> In your case, since you want to handle diverging paths in the same way as normal paths, I think you could use `Value.normal_return` as the diverging value (and get rid of the `div` field of your record).

I considered it but  it would be more conservative in some cases than tracking divergence as a separate component: 
```
let[@zero_alloc][@inline never] f y =
  if y > 0 then raise (Exn_int y)
  else
    while true do
      Sys.opaque_identity ()
    done

let[@zero_alloc] g x y=
   f y;
   (x,x)
```
The summary for `f` is `{ nor = Bot; exn = Top; div = Safe }`  and in particular `f` does not return normally, so we know that the allocation in `g` after the call to `f` is not reachable, and the summary for `g` is `{ nor = Bot; exn = Top; div = Safe}` which satisfies the relaxed annotation. If we do not distinguish normal return from divergence, the summary for `f` would be `{ nor = Safe; exn = Top } and the summary for `g` would be `{ nor = Top; exn = Top }`. It's a pretty contrived example though.

**Comment by xclerc** [comment]
The summaries I get for both `f` and `g` seem
to be `{ nor = Top; exn = Top; div = Top }`.

**Comment by xclerc** [comment]
I am still not sure about the semantics of `div`.
For instance, `let[@inline never] rec forever () : unit = forever ()` gives:
- closure: `{ nor=bot; exn=bot; div=bot }`;
- flambda:  `{ nor=bot; exn=bot; div=bot }`;
- flambda2: `{ nor=bot; exn=bot; div=safe }`.
I see how it happens, but am basically afraid it could create problems
down the line.

**Comment by gretay-js** [comment]
> I am still not sure about the semantics of `div`. For instance, `let[@inline never] rec forever () : unit = forever ()` gives:
> 
> * closure: `{ nor=bot; exn=bot; div=bot }`;
> * flambda:  `{ nor=bot; exn=bot; div=bot }`;
> * flambda2: `{ nor=bot; exn=bot; div=safe }`.
>   I see how it happens, but am basically afraid it could create problems
>   down the line.

Are you referring to how the different middle-end optimizations affect the result of this analysis? It is hard to define the meaning of the analysis components at the level of source code.

**Comment by xclerc** [comment]
OK, but isn't is weird to have summaries
with all three components set to `Bot`?

**Comment by xclerc** [comment]
I think it comes from the way `callees`/`callers`
are processed, but with no knowledge of the
implementation I think it is difficult to see why
it is rejected.

**Comment by lthls** [comment]
> The reason I switched to using Backward analysis is to compute backward reachability from normal and exceptional return.

I'm not sure I understand. We have a graph with one entry node `I` and two exit nodes `O_N` and `O_E` (forgetting diverging paths for now). What we are interested in is what happens on paths from `I` to `O_N`, and on paths from `I` to `O_E`. With a forwards analysis you just propagate data from `I` along the edges, and read the results for `O_N` and `O_E` separately. With a backwards analysis you have to do two different propagations (one from `O_N` and one from `O_E`), and your version uses a record with two independent fields for that. That should work too, but looks more complicated to me than the forwards version.
But I'm now convinced that we wouldn't gain much by switching to a forwards dataflow analysis, so let's put this particular line of thought to rest.

I have looked at the handling of diverging paths, and I am convinced that your two-phase approach is wrong.
Here is an example:
```ocaml
let f cond =
  while true do
    allocate();
    if Sys.opaque_identity cond then raise Not_found
  done
```
Here I'm expecting the analysis to return `{ nor = Bot; exn = Top; div = Top }`. But if I understand correctly, your first pass will mark the loop as reached (as it's backwards reachable from the exceptional return) and won't introduce diverging values on the relevant loop labels. So I believe your code returns `{ nor = Bot; exn = Top; div = Bot }` instead of the value I expect.
I believe that the fix is easy: just do a single call to `D.analyze` with `init_lbl` set to `Value.diverges` unconditionally. Since `init_lbl` only gets called when looking at back-edges, that should give you the correct value for all three paths in a single analysis.
And I suspect that the weird handling of `raise_notrace` can be reverted back once you've done that.

Finally, I believe that recursive self calls (and probably future calls in general, though I haven't checked) are treated unsoundly too. Here is an example:
```ocaml
let rec g x =
  if Sys.opaque_identity true then
    try g x with _ -> ()
  else raise (Failure x)
```
The current result for this function is `{ nor=bot; exn=top; div=bot }`. The `div` part is clearly unsound, but even worse, the `nor` part is unsound. This is because we only go through the body of `g` once, assuming that `g` has either `Value.bot` (self case) or `Value.safe` (if we used the future function path), and so infer that the normal path is either `bot` or `safe`, and the exception path is `top`. But if we knew that the exception path was `top`, then we would have inferred `top` on the normal path too.
This shows the need for a fixpoint for soundness of recursive functions. The alternative is to used `Value.top` instead of `Value.safe` for future functions, and I suspect this will lead to an unacceptable number of false positives.

**Comment by gretay-js** [comment]
@xclerc @lthls I believe that I addresses all the comments above, as discussed offline.  
Main changes:
- Single backward data flow analysis pass, setting the initial abstract value of `Iexit` targets to Safe instead of Bot.
- Update the meaning of Relaxed annotation to ignore `div` component.
- Conservative handling of self calls 
- Conservative use of summaries of unresolved dependencies
- Improved conservative indirect calls handling
- Added tests from this discussion 

I have a collection of followup PRs, including additional attributes and a proposal of a domain for more precise tracking of interproc dependencies (the outer fixpoint between function summaries).

**Comment by xclerc** [comment]
I have not really started the review, but noticed the following example passes:
```
let[@inline never] f x = if x > 0 then (Printf.eprintf "%d\n%!" x; raise Not_found)
let[@zero_alloc] g x = f x; (x, x)
let[@zero_alloc] g x = x
```
Commenting out the last line makes it fail on line 2, as expected.

(This is with flambda2 / [95a3310](https://github.com/ocaml-flambda/flambda-backend/pull/1157/commits/95a331046260d607b318c4089cf87224bd124d0b))

**Comment by lthls** [comment]
> I have not really started the review, but noticed the following example passes:
> 
> ```
> let[@inline never] f x = if x > 0 then (Printf.eprintf "%d\n%!" x; raise Not_found)
> let[@zero_alloc] g x = f x; (x, x)
> let[@zero_alloc] g x = x
> ```
> 
> Commenting out the last line makes it fail on line 2, as expected.
> 
> (This is with flambda2 / [95a3310](https://github.com/ocaml-flambda/flambda-backend/pull/1157/commits/95a331046260d607b318c4089cf87224bd124d0b))

I suspect that with flambda2, both `f` and the first version of `g` are removed because they're unused (`f` might remain if it is exported in the interface, either because it appears in the `.mli` or if there is no `.mli` file).
So there's no way to know in the backend that there was another function to check.

I don't know if this is a problem or not, but if you think it is I suggest opening another issue so that we can discuss what to do (it's completely unrelated to this PR as far as I can tell).
I think that if you insert `ignore (Sys.opaque_identity g)` in the body of the second version of `g` (or at toplevel between the definitions of `g`) then the first version will be kept alive and the check will trigger. It's a pain to do for the users, but if it works we can do it automatically in the compiler too.

**Comment by xclerc** [comment]
What you describe is probably at play, but I came to it through
```
let[@inline never] f x = if x > 0 then (Printf.eprintf "%d\n%!" x; raise Not_found)
let[@zero_alloc] g x = f x; (x, x)
let h x = g x
let[@zero_alloc] g x = x
```
which is puzzling. Fortunately, either marking the first `g` as
inline never of `h` as `zero_alloc` triggers the expected failure.

**Comment by xclerc** [comment]
Oh, indeed I see the chain of events; it just fells not very user-friendly.

**Comment by xclerc** [comment]
I am not sure I understand why `Icheckbound` is
treated as "simply" allocating. The two functions
below are morally equivalent, but only the second
one passes the test:
```
let[@zero_alloc] f a i : int = a.(i)

let[@zero_alloc] g a i : int =
  if i >= 0 && i < Array.length a then
    Array.unsafe_get a i
  else
    raise (Invalid_argument "index out of bound")
```

(This time with closure, flambda2 avoiding `Icheckbound`.)

**Comment by xclerc** [comment]
In fact, it may very well be the case that `f` never allocates.

**Comment by gretay-js** [comment]
> I am not sure I understand why `Icheckbound` is treated as "simply" allocating. 

This was too conservative, here is a more precise transformer for it.

File: backend/checkmach.ml [hunk 9]
```diff
@@ -27,454 +27,809 @@
...
+    | None -> None
+    | Some ({ nor; exn; div } : Checks.value) ->
+      Some
+        { Value.nor = decode_return nor;
+          exn = decode_return exn;
+          div = decode_diverge div
+        }
+
+  let transform_specific s ~next ~exn:_ =
+    if Arch.operation_allocates s then Value.transform next else next
```
Comment (xclerc) [inline backend/checkmach.ml:817]: I am surprised this does not also use
`Arch.operation_can_raise`. Arguably,
that will not make a difference at the
moment on amd64...

File: backend/checkmach.ml [hunk 10]
```diff
@@ -27,454 +27,809 @@
...
+       diverge (non-terminating loops that do not reach normal or exceptional
+       return). All loops must go through an (Iexit label) instruction or a
+       recursive function call. If (Iexit label) is not backward reachable from
+       the function's Normal or Exceptional Return, either the loop diverges or
+       the Iexit instruction is not reachable from function entry.
+
+       To check divergent loops, the initial value of "div" component of all
+       Iexit labels is set to "Safe" instead of "Bot". This is conservative with
+       respect to non-recursive Icatch. *)
+    D.analyze ~exnescape:Value.exn_escape ~init_lbl:Value.diverges ~transfer
```
Comment (xclerc) [inline backend/checkmach.ml:732]: `init_lbl` will set `div` to `Safe` for all labels,
including `Iexit` ones, right?

Comment (gretay-js) [reply backend/checkmach.ml:732]: Yes, including `Iexit` labels used by non-recursive Icatch handlers.

Comment (gretay-js) [reply backend/checkmach.ml:732]: I may have misunderstoond your question. Are you asking about labels used for `Itrywith` handlers rather than `Icatch` handlers? I think with `flambda2` these will also get `div` set to `Safe` instead of `Bot` by default, unfortunately, in the current version.

File: backend/checkmach.ml (hunk 9 above)
Comment (gretay-js) [reply backend/checkmach.ml]: See proposed fix in [commit 5ad2314](https://github.com/ocaml-flambda/flambda-backend/pull/1157/commits/5ad23142261500b52ae605c197854ebad6efe8a7). I'm assuming that specific operations cannot diverge.

File: middle_end/flambda2/terms/check_attribute.ml [hunk 11]
```diff
@@ -13,7 +13,7 @@
 module Property = struct
   type t = Noalloc
 
-  let to_string = function Noalloc -> "noalloc"
+  let print ppf = function Noalloc -> Format.fprintf ppf "alloc"
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply middle_end/flambda2/terms/check_attribute.ml]: - renamed Lambda.property.Noalloc to Zero_alloc, and updated downstream from there, hopefully catching most of these. 
- renamed `-alloc-check` to `-zero-alloc-check` flag and the corresponding arg.

File: backend/printcmm.ml [hunk 12]
```diff
@@ -361,15 +361,18 @@ and sequence ppf = function
 
 and expression ppf e = fprintf ppf "%a" expr e
 
-let property : Cmm.property -> string = function
-    | Noalloc -> "noalloc"
+let property_to_string : Cmm.property -> string = function
+  | Noalloc -> "zero_alloc"
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/printcmm.ml]: Done.

File: backend/checks.ml [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+  t.enabled <- true
+
+module Raw = struct
+  type entries = (string * bool) list
+
+  type r =
+    { nor : entries;
+      exn : entries;
+      div : entries
+    }
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/checks.ml:68]: Yes, the current representation can be significantly optimized. I opened https://github.com/ocaml-flambda/flambda-backend/issues/1280 to implement it.
Currently, on the compiler build itself and the distribution, the overhead is under 6% of the size of .cmx files. The check is off by default outside of compiler libraries.

File: backend/checks.ml [hunk 14]
```diff
@@ -0,0 +1,110 @@
+module String = Misc.Stdlib.String
+
+(* CR gyorsh: refactor repetititve code. *)
+(* CR gyorsh: Add [t] per analysis when at least one more analysis is
+   implemented *)
+type t =
+  { mutable nor : bool String.Map.t;
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/checks.ml:7]: It was useful at some point when the maps where sparse, but with `div` is probably worse. I would like to leave optimizations for a separate PR, unless there is a (soundness) bug.

File: backend/checkmach.ml (hunk 10 above)
Comment (gretay-js) [reply backend/checkmach.ml]: Updated comment.

File: backend/checkmach.ml [hunk 15]
```diff
@@ -27,454 +27,813 @@
...
 module type Spec = sig
-  (** Which check is it? Used in error messages. *)
-  val name : string
-
   (** Is the check enabled? *)
   val enabled : unit -> bool
 
-  (** Record that the callee passed the check. *)
-  val add_callee : string -> Cmx_format.checks -> unit
+  (** [get_value_opt f] returns the value recorded for function [f] in [Compenv], either
```
[the remark this answers is not in the document]
Comment (gretay-js) [reply backend/checkmach.ml]: Applied.

**Review by gretay-js — COMMENTED** [review]
I think I addressed all the comments so far.

File: tests/backend/checkmach/s.ml [hunk 16]
```diff
@@ -1,2 +1,8 @@
-let[@noalloc][@inline never] foo n m = n + m
+let[@zero_alloc strict][@inline never] foo n m = n + m
 let[@inline never] bar n m = (n, m)
```
Comment (xclerc) [inline tests/backend/checkmach/s.ml:2]:
```suggestion
```

**Comment by xclerc** [comment]
The following is really a corner case, and quite
possibly not one we want to support:
```
exception E
type t = { x : int; mutable y : int }
let[@zero_alloc] f {x; y} () = Printf.eprintf "%d\n%!" (y); raise E
```
It fails because the `zero_alloc` annotation is
left on a function which end up simply calling
the helper, which is the one unconditionally
raising.

**Comment by xclerc** [comment]
(Actually, proper support for "forward" references
will quite probably be enough to support this case.)

**Comment by gretay-js** [comment]
> (Actually, proper support for "forward" references will quite probably be enough to support this case.)

I don't think so, there is no call to the helper function. It's a bug - the annotation should not  be on the helper function.
It feels similar to https://github.com/ocaml-flambda/flambda-backend/issues/1273  in the sense that the optimizer inlines/lifts some functions.

**Comment by xclerc** [comment]
Oh, indeed I misread the output of `-dsel`: there
is a _reference_ to the helper function, but not a
_call_ to it.

**Comment by xclerc** [comment]
(And to be clear, as far as I can see the annotation
stays on the original function, which happens to be
essentially a stub eventually.)

**Comment by lthls** [comment]
> The following is really a corner case, and quite possibly not one we want to support:
> 
> ```
> exception E
> type t = { x : int; mutable y : int }
> let[@zero_alloc] f {x; y} () = Printf.eprintf "%d\n%!" (y); raise E
> ```
> 
> It fails because the `zero_alloc` annotation is left on a function which end up simply calling the helper, which is the one unconditionally raising.

I think this is one of those cases where mutable patterns cause functions to have different arities than expected. In this case, we end up with `let[@zero_alloc] f {x; y} = (fun () -> Printf.eprintf "%d\n%!" (y); raise E)`, which is allocating as it returns a closure containing one of its arguments.
 You can have a look at ocaml/RFCs#32, which describes a small change to the language that would solve this problem (or you could talk with @stedolan directly, he might like having another concrete example where his proposal would fix bugs).

File: tests/backend/checkmach/dune [hunk 17]
```diff
@@ -1,14 +1,17 @@
+;; CR-someday gyorsh: refactor this file using something like cinaps to
```
Comment (xclerc) [inline tests/backend/checkmach/dune:1]: Cinaps would pull a dependency, but maybe
the technique used in https://github.com/ocaml-flambda/flambda-backend/pull/1244 can be applied
here too.

File: backend/checks.ml (hunk 13 above)
[the remark this answers is not in the document]
Comment (xclerc) [reply backend/checks.ml:68]: 6% percent is indeed not a big deal
at this point. By the way, do you have
the number for running the analysis?
(_i.e._ `Profile.record`)

**Comment by xclerc** [comment]
Two usability quibbles, for a follow-up PR:
- it would be nice to give to the user the
  location of a violation, rather than only
  the fail/pass status;
- possibly also keep the name of the
  function as it appears in the source for
  the error messages.

File: backend/checkmach.ml [hunk 18]
```diff
@@ -27,454 +27,813 @@
...
-    | Ialloc { mode = Alloc_heap; _ } -> report_fail t "allocation" dbg
-    | Ipoll _ ->
-      (* Polling points may trigger finalisers, signal handlers or memprof
-         callbacks, but any allocations done therein don't count towards the
-         calling function being "allocating". *)
-      ()
+      (* does not allocate even when it raises because checkbound exception is
+         static. *)
+      transform t ~next ~exn ~effect:Value.safe "checkbound" dbg
+    | Ipoll _ (* CR gyorsh: poll points are considered allocations. *)
```
Comment (xclerc) [inline backend/checkmach.ml:658]: This pattern should either be moved
to the `Alloc_heap` case, or have code
that ensure we will revisit it when poll
points are generated (_i.e._ `assert false`
or `Misc.fatal`).

File: file_formats/cmx_format.mli [hunk 19]
```diff
@@ -83,7 +75,7 @@ type unit_infos_raw =
     uir_imports_cmx: Import_info.t array;
     uir_generic_fns: generic_fns;
     uir_export_info: export_info_raw;
-    uir_checks: checks;
+    uir_checks: Checks.Raw.t;
```
Comment (xclerc) [inline file_formats/cmx_format.mli:78]: (This probably means we should
bump the magic number.)

File: tests/backend/checkmach/dune (hunk 17 above)
Comment (gretay-js) [reply tests/backend/checkmach/dune:1]: Ah, yes, that's a good idea. I can start by adding a generator for dune these rules, and then simplify the dune file further using the "include files from sexp"  feature when we switch to `dune 3.5`.

File: backend/checkmach.ml (hunk 18 above)
Comment (gretay-js) [reply backend/checkmach.ml]: Fixed.

**Review by xclerc — APPROVED** [review]

**Comment by gretay-js** [comment]
@lthls do you have any other comments about this version? (A follow up PR will do the outer fixpoint on function summaries, lifted from `Value.t` to essentially `Env.t->Value.t`.)

**Comment by lthls** [comment]
I don't have anything new to add, for me this PR can go forward.
